e2e/nomostest/testresourcegroup/resourcegroup.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testresourcegroup
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"kpt.dev/configsync/e2e/nomostest/testkubeclient"
+	"kpt.dev/configsync/pkg/api/configmanagement"
+	"kpt.dev/resourcegroup/apis/kpt.dev/v1alpha1"
+	"kpt.dev/resourcegroup/controllers/resourcegroup"
+	"sigs.k8s.io/cli-utils/pkg/common"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// NotOwnedMessage is the status message for a resource which does not have
+	// an inventory id.
+	NotOwnedMessage = "This object is not owned by any inventory object. The status for the current object may not reflect the specification for it in current ResourceGroup."
+)
+
+// NewKptResourceGroup creates a new ResourceGroup object
+func NewKptResourceGroup(nn types.NamespacedName, id string) *v1alpha1.ResourceGroup {
+	rg := &v1alpha1.ResourceGroup{}
+	rg.SetNamespace(nn.Namespace)
+	rg.SetName(nn.Name)
+	if id != "" {
+		rg.SetLabels(map[string]string{
+			common.InventoryLabel: id,
+		})
+	}
+	return rg
+}
+
+// AggregateResourceStatus is a helper function to generate an array of ResourceStatus
+// from ObjMetadata
+func AggregateResourceStatus(resources []v1alpha1.ObjMetadata, s v1alpha1.Status, sourceHash string) []v1alpha1.ResourceStatus {
+	resourceStatus := []v1alpha1.ResourceStatus{}
+	for _, resource := range resources {
+		resourceStatus = append(resourceStatus, v1alpha1.ResourceStatus{
+			ObjMetadata: resource,
+			Status:      s,
+			SourceHash:  sourceHash,
+		})
+	}
+	return resourceStatus
+}
+
+// InitialResourceStatus returns the initial/default ResourceStatus of a
+// ResourceGroup object.
+func InitialResourceStatus() v1alpha1.ResourceGroupStatus {
```

Comment:
I'm confused. Why is the initial status FinishReconciling?  I would have expected the initial status to just not includes these conditions at all.

Reply:
Seems to be a quirk of the resource-group controller. Looks like it comes from https://github.com/GoogleContainerTools/kpt-resource-group/blob/main/controllers/resourcegroup/resourcegroup_controller.go#L216

Reply:
Oh, I see. It's not the initial status so much as the success status set after all resource statuses have been computed without stalling. But if there's no resources, then it reconciles immediately.

How about `EmptyStatus()` with a bit more comment?

Reply:
Done.